core/dbt/node_runners.py

```diff
@@ -467,11 +333,96 @@ def execute(self, model, manifest):
         return RunModelResult(model, status=result.status)
 
 
-class TestRunner(CompileRunner):
+class FreshnessRunner(BaseRunner):
+    def on_skip(self):
+        raise dbt.exceptions.RuntimeException(
+            'Freshness: nodes cannot be skipped!'
+        )
+
+    def before_execute(self):
+        description = 'freshness of {0.source_name}.{0.name}'.format(self.node)
+        dbt.ui.printer.print_start_line(description, self.node_index,
+                                        self.num_nodes)
+
+    def after_execute(self, result):
+        dbt.ui.printer.print_freshness_result_line(result,
+                                                   self.node_index,
+                                                   self.num_nodes)
+
+    @staticmethod
+    def _freshness_delta(freshness):
+        # timedelta makes this convenient!
+        kwargs = {freshness['period']+'s': freshness['count']}
+        return timedelta(**kwargs).total_seconds()
+
+    def _calculate_status(self, target_freshness, freshness):
+        """Calculate the status of a run.
+
+        :param dict target_freshness: The target freshness dictionary. It must
+            match the freshness spec.
+        :param timedelta freshness: The actual freshness of the data, as
+            calculated from the database's timestamps
+        """
+        warn_after = self._freshness_delta(target_freshness['warn_after'])
+        error_after = self._freshness_delta(target_freshness['error_after'])
+        # if freshness > warn_after > error_after, you'll get an error, not a
+        # warning
+        if freshness > error_after:
+            return 'error'
+        elif freshness > warn_after:
+            return 'warn'
+        else:
+            return 'pass'
+
+    def _build_run_result(self, node, start_time, error, status, timing_info,
+                          skip=False, failed=None):
+        execution_time = time.time() - start_time
+        thread_id = threading.current_thread().name
+        timing = [t.serialize() for t in timing_info]
+        if status is not None:
+            status = status.lower()
+        return PartialResult(
+            node=node,
+            status=status,
+            error=error,
+            execution_time=execution_time,
+            thread_id=thread_id,
+            timing=timing
+        )
 
-    def raise_on_first_error(self):
-        return False
+    def from_run_result(self, result, start_time, timing_info):
+        result.execution_time = (time.time() - start_time)
+        result.timing.extend(t.serialize() for t in timing_info)
+        return result
+
+    def execute(self, compiled_node, manifest):
+        # given a Source, calculate its fresnhess.
+        freshness = self.adapter.calculate_freshness(
```

Comment:
i couldn't track down exactly where this happens, but i think we try to calc freshness for sources that don't have a `freshness` key specified. With this schema.yml:

```
version: 2

sources:
  - name: snowplow
    loader: snowpipe
    tables:
      - name: event
        sql_table_name: raw.snowplow.event
        loaded_at_field: collector_tstamp
```

i see:
```
21:27:32 | 1 of 1 START freshness of snowplow.event............................. [RUN]
Unhandled error while executing source.dbt_metrics.snowplow.event
'warn_after'
21:27:34 | 1 of 1 ERROR......................................................... [ERROR in 1.74s]
```

Reply:
similarly, it looks like `warn_after` and `error_after` are both required by the code in here, but i think each is optional per the contract? It's totally reasonable to specify one of these and not the other

Reply:
I'll poke around a bit, we should not be getting that node back from the selector at all.

Requiring both `warn_after` and `error_after` was my understanding of the initial feature but it's easy enough to change.

Reply:
Ok, I fixed this by checking if `self.freshness` as truthy instead of a not-none value, which handles the empty dict case, as it should. it also handles the absence of either one of the values.

Now you'll get the `WARNING: Nothing to do. Try checking your model configs and model specification args` you'd expect.